multiaddr.go

```diff
@@ -38,39 +53,37 @@ func NewMultiaddrBytes(b []byte) (a Multiaddr, err error) {
 	}()
 
 	if err := validateBytes(b); err != nil {
-		return nil, err
+		return Multiaddr{}, err
 	}
 
-	return &multiaddr{bytes: b}, nil
+	return Multiaddr{bytes: bstr(b)}, nil
 }
 
-// Equal tests whether two multiaddrs are equal
-func (m *multiaddr) Equal(m2 Multiaddr) bool {
-	return bytes.Equal(m.bytes, m2.Bytes())
+// Equal returns whether two Multiaddrs are exactly equal
+func (m Multiaddr) Equal(m2 Multiaddr) bool {
+	return m.bytes == m2.bytes
 }
 
 // Bytes returns the []byte representation of this Multiaddr
-func (m *multiaddr) Bytes() []byte {
-	// consider returning copy to prevent changing underneath us?
-	cpy := make([]byte, len(m.bytes))
-	copy(cpy, m.bytes)
-	return cpy
+func (m Multiaddr) Bytes() []byte {
+	return []byte(m.bytes)
 }
 
-// String returns the string representation of a Multiaddr
-func (m *multiaddr) String() string {
-	s, err := bytesToString(m.bytes)
+// String returns the string representation of this Multiaddr
+// (may panic if internal state is corrupted)
+func (m Multiaddr) String() string {
+	s, err := bytesToString(m.Bytes())
```

Comment:
if we're going down this path, we might has well have `bstrToString`

Reply:
We'd just have to call `Bytes()` inside it as we can't extract varints from strings.